Updating binary file formats

paulromano opened the pull request:
This PR makes updates in almost all our binary file formats, and especially for the summary and statepoint files there are significant changes. For the summary file:

- Some metadata is no longer written (number of particles, number of batches, etc.)
- `n_cells`, `n_surfaces`, `n_materials` are now attributes rather than datasets
- The array indices used within the OpenMC solver (`index` datasets) are not written anymore. Elsewhere they are changed to user IDs (e.g., list of cells for each universe, lattice outer universe.)
- Tally names were moved to the statepoint file where all the rest of the tally data is.
- The filetype, file format version, OpenMC version, git SHA1 hash, and time stamp are written as attributes to the root group.

Along with this, the `openmc.Summary()` class was heavily modified:

- The `openmc_geometry` attribute is now just `geometry`. The `opencg_geometry` attribute has been removed. (the same change was made for the `mgxs.Library` class)
- A `materials` attribute was added that is of type `openmc.Materials`.
- Reading of surfaces was moved to a new `openmc.Surface.from_hdf5` staticmethod.
- Many undocumented attributes were removed. All the remaining ones are now documented in the class docstring.
- Many of the `get_*_by_id` methods were removed. I'd be open to adding these back, but perhaps not in the `Summary` class (rather in `Geometry`, `Materials`, etc.)


For the statepoint file, the following changes were made:

- The filetype, file format version, OpenMC version, git SHA1 hash, and time stamp are written as attributes to the root group.
- `run_CE` is now `energy_mode` and is a string.
- Several datasets were turned into attributes.
- `gen_per_batch` is now `generations_per_batch`
- Internal array indices were removed. User IDs for tallies and meshes are now written in an `ids` attribute.
- The `to_statepoint` type-bound procedures for filters now write user IDs rather than internal array indices. The one exception is surface filters (that are only used for mesh surface currents).

Corresponding changes were also made in the StatePoint class. Note that the only extra information you get with the `link_with_summary` method now is distribcell paths -- everything else appears by default.

All HDF5 formats have been modified to have `filetype` and `version` attributes. Each file format has a two digit version number now, addressing #766. The Python API now uses a new function, `checkvalue.check_filetype_vesion`, for each corresponding class. The documentation for all file formats has been updated (note that there were a number of things that were out of date in the existing documentation that are now fixed).

While I was updating the voxel file format, I also decided to simplify the logic for writing PPM files (removing ppmlib.F90 altogether) and added OpenMP parallelization as well. On my laptop, generating plots for the BEAVRS model used to take about 73 s and it now takes 22 s (about what you'd expect for a hyperthreaded dual-core processor). I had to remove the progress bar because multiple threads are making progress simultaneously.

One other plot related change -- the default file names are now plot_N.ppm rather than N_plot.ppm to be consistent with other files.

@wbinventor I'm requesting your review because you are probably most familiar with what's in these files.

wbinventor commented:
I'll at least start revising this today. Before I begin, I must ask is there is a design philosophy you aim to realize through this assortment of changes? If so, I'll try to review this through that lens.

paulromano commented:
Not sure I would call it a philosophy, but a few basic considerations that led to these changes:

- Every HDF5 file should have a `filetype` and `version` attribute, stored in a consistent manner. The only exception now is the WMP file format which has a string version rather than a pair of ints.
- The names of datasets in file and attributes on Python classes should match what we use for the inputs. To give two examples: I renamed `run_CE` to `energy_mode` because the latter is what we use in the `Settings` class. I also renamed `gen_per_batch` to `generations_per_batch` for the same reason.
- The `Settings` class should essentially just capture the geometry and materials information.
- I wanted to move some metadata (things like `n_cells`) into attributes rather than datasets because 1) that's where metadata belongs and 2) attributes take much less memory. Admittedly, I could have moved a lot more stuff into attributes, but I didn't want to *completely* change the format.

One question for you -- do you (or others) use those `get_*_by_id` methods often enough to keep them around in the API? If so, I can add them back in the `Geometry` and `Materials` classes.

paulromano commented on docs/source/io_formats/statepoint.rst:
```diff
@@ -4,354 +4,176 @@
+             - **date_and_time** (*char[]*) -- Date and time the summary was
+               written.
+             - **path** (*char[]*) -- Path to directory containing input files.
+             - **cmfd_on** (*int*) -- Flag indicating whether CMFD is on (1) or
```
Booleans do not exist in [HDF5](http://ftp.hdfgroup.org/hdf5-quest.html#bool).

paulromano commented on docs/source/io_formats/summary.rst:
```diff
@@ -4,255 +4,131 @@
+           - **universe** (*int*) -- Universe assigned to the cell. If none is
+             specified, the default universe (0) is assigned.
+           - **fill_type** (*char[]*) -- Type of fill for the cell. Can be
+             'normal', 'universe', or 'lattice'.
```
That has been bothering me too. I think I'll change it to material as part of this PR if you're ok with it.

wbinventor replied:
Sounds good.

paulromano commented on docs/source/pythonapi/examples/mgxs-part-ii.ipynb:
```diff
@@ -46,14 +46,15 @@
     "import numpy as np\n",
     "import matplotlib.pyplot as plt\n",
-    "import seaborn as sns\n",
+    "plt.style.use('seaborn-dark')\n",
```
Yes, matplotlib 1.5+ I believe.

wbinventor replied:
Sounds like progress.

paulromano commented on docs/source/pythonapi/examples/mgxs-part-ii.ipynb:
```diff
@@ -483,85 +484,93 @@
+      "       79/1    1.22125    1.22334 +/- 0.00187\n",
+      "       80/1    1.23793    1.22355 +/- 0.00186\n",
+      "       81/1    1.24238    1.22382 +/- 0.00185\n",
+      "       82/1    1.23493    1.22397 +/- 0.00183\n",
```
That is because of #786.

paulromano commented on openmc/checkvalue.py:
```diff
@@ -239,6 +239,39 @@ def check_greater_than(name, value, minimum, equality=False):
+
+        # Check version
+        if this_version[0] != expected_version:
+            raise IOError('Statepoint file has a version of {} which is not '
```
Yes, thanks for catching that.

paulromano commented on openmc/statepoint.py:
```diff
@@ -147,12 +138,9 @@ def __init__(self, filename, autolink=True):
                     vol = openmc.VolumeCalculation.from_hdf5(path_i)
                     self.add_volume_information(vol)
 
-    def close(self):
```
h5py automatically closes a file when a `h5py.File` object goes out of scope. Thus, when a statepoint goes out of scope, the file will also close.

paulromano commented on openmc/statepoint.py:
```diff
@@ -478,21 +418,20 @@ def tally_derivatives(self):
                     else:
-                        raise RuntimeError('Unrecognized tally differential '
-                                           'variable')
+                        raise ValueError('Unrecognized tally differential '
```
Actually, the setter for `TallyDerivative.variable` already handles this so I'm just going to remove this exception.

paulromano commented on openmc/summary.py:
```diff
@@ -135,365 +117,245 @@ def _read_materials(self):
-        for key in self._f['geometry/lattices'].keys():
-            if key == 'n_lattices':
-                continue
+    def _read_lattices(self, universes):
```
I did it for the cases where it's obvious that there's no extra information needed. For example, to have a `Cell.from_hdf5` method, you also potentially need material and universe information (to assign the fill). However, at the time we're reading cells, we don't have universe information yet. All that information gets combined in the `_finalize_geometry` method. I can revisit this though and see if it might be feasible by changing the order of things.

paulromano commented on openmc/summary.py:
```diff
@@ -567,92 +403,8 @@ def get_material_by_id(self, material_id):
                 return material
 
         return None
-
```
I propose that the `get_all_*` method always return a dictionary mapping IDs to things (unlike what `Geometry.get_all_*` currently does). If the user really wants a list of things, they can just do `list(__.get_all_things().values())`. To get a cell by ID would be `geometry.get_all_cells()[my_id]`.

Alternatively, we could always return a list of objects. That way, we could leave it up to the user what kind of dictionary they want, e.g.
```Python
cell_by_name = {c.name: c for c in geometry.get_all_cells()}
```

So, whatever we do, we should make sure that `Geometry.get_all_cells()` and `Universe.get_all_cells()` are consistent since they are not currently. Let me know what you think.

wbinventor commented:
I'm all about returning a dictionary mapping IDs to objects. Let's just
stick to that across the board! Who knows why I decided to return lists, it
makes no sense to me now.
On Wed, Feb 22, 2017 at 6:28 PM Paul Romano <notifications@github.com>
wrote:

> *@paulromano* commented on this pull request.
> ------------------------------
>
> In openmc/summary.py
> <https://github.com/mit-crpg/openmc/pull/794#discussion_r102602136>:
>
> >              if material.id == material_id:
>                  return material
>
>          return None
> -
>
> I propose that the get_all_* method always return a dictionary mapping
> IDs to things (unlike what Geometry.get_all_* currently does). If the
> user really wants a list of things, they can just do
> list(__.get_all_things().values()). To get a cell by ID would be
> geometry.get_all_cells()[my_id].
>
> Alternatively, we could always return a list of objects. That way, we
> could leave it up to the user what kind of dictionary they want, e.g.
>
> cell_by_name = {c.name: c for c in geometry.get_all_cells()}
>
> So, whatever we do, we should make sure that Geometry.get_all_cells() and
> Universe.get_all_cells() are consistent since they are not currently. Let
> me know what you think.
>
> —
> You are receiving this because you were assigned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/794#discussion_r102602136>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAMyVCncWmXb_dU_IhKoi8UOjG3ChKy2ks5rfMSMgaJpZM4MIKIe>
> .
>

paulromano commented:
Ok, I think all comments have been addressed now.

wbinventor commented on openmc/cell.py:
```diff
@@ -563,7 +563,7 @@ def create_surface_elements(node, element):
                     path = "./surface[@id='{}']".format(node.surface.id)
                     if xml_element.find(path) is None:
-                        xml_element.append(node.surface.create_xml_subelement())
+                        xml_element.append(node.surface.to_xml_element())
```
I like the shorter `to_xml_*()` name better. Perhaps everything should be converted to this naming convention in the future.

paulromano replied:
Yes, I wanted to do it with Cell, Lattice, and Universe too but they are a bit more complicated so I held off for now.

wbinventor commented on docs/source/io_formats/summary.rst:
```diff
@@ -31,7 +31,7 @@ The current version of the summary file format is 5.0.
              specified, the default universe (0) is assigned.
            - **fill_type** (*char[]*) -- Type of fill for the cell. Can be
-             'normal', 'universe', or 'lattice'.
+             'material', 'universe', or 'lattice'.
```
Is this called `'material'` or `'distribmat'` for a distrimat fill?

wbinventor commented on src/constants.F90:
```diff
@@ -107,11 +107,11 @@ module constants
-       CELL_NORMAL  = 1, & ! Cell with a specified material
-       CELL_FILL    = 2, & ! Cell filled by a separate universe
-       CELL_LATTICE = 3    ! Cell filled with a lattice
+       FILL_MATERIAL = 1, & ! Cell with a specified material
```
Much cleaner

wbinventor commented on openmc/cell.py:
```diff
@@ -485,20 +485,23 @@ def get_all_materials(self):
-        cells = self.get_all_cells()
-        for cell in cells.values():
-            materials.update(cell.get_all_materials())
+        elif self.fill_type == 'distribmat':
```
I think that seeing this made me wonder if `'distribmat'` should be considered a fill type in our file format specifications.

paulromano replied:
That's a good point. In fact, considering it as a separate fill type might simplify things on the Fortran side as well. Let me look into that.

wbinventor commented on openmc/cell.py:
```diff
@@ -485,20 +485,23 @@ def get_all_materials(self):
-            materials.update(cell.get_all_materials())
+        elif self.fill_type == 'distribmat':
+            for m in self.fill:
+                if m is not None:
```
Any reason `m` could be `None`?

paulromano replied:
Yes, in theory, one of the distributed instances might be a void material, in which case `m is None`

wbinventor commented on openmc/geometry.py:
```diff
@@ -151,123 +151,98 @@ def get_cell_instance(self, path):
 
     def get_all_material_universes(self):
-        """Return all universes composed of at least one non-fill cell
+        """Return all universes having at least one non-fill cell
```
"at least one non-fill cell" How about "at least one material-filled cell"?

wbinventor replied:
I know that I'm probably asking you to correct a confusing comment that I probably once wrote :-)

paulromano replied:
Haha, I also thought it sounded awkward. I'll fix that.

wbinventor commented on openmc/geometry.py:
```diff
@@ -151,123 +151,98 @@ def get_cell_instance(self, path):
-            Universes with non-fill cells
+        collections.OrderedDict
+            Dictionary mapping universe IDs to :class:`openmc.Universe`
+            instances with non-fill cells
```
Ditto, "non-fill" -> "material-filled"

wbinventor commented on openmc/geometry.py:
```diff
@@ -151,123 +151,98 @@ def get_cell_instance(self, path):
-            for cell in cells:
+        for universe in self.get_all_universes():
+            for cell in universe.cells:
                 if cell.fill_type in ('material', 'distribmat', 'void'):
```
It might be nice to clarify in the docstring that this method treats `'void'` as a "material".

wbinventor commented on openmc/mgxs/library.py:
```diff
@@ -192,11 +192,11 @@ def domain_type(self):
         if self._domains == 'all':
             if self.domain_type == 'material':
-                return self.geometry.get_all_materials()
+                return list(self.geometry.get_all_materials().values())
```
Just for sanity's sake, have you re-run our Jupyter notebooks with the very latest changes?

paulromano replied:
I re-ran them a few commits ago, but I'll double check again just for good measure.

wbinventor commented:
I'm a big fan of this PR! Just made a few final comments.

paulromano commented:
@wbinventor Final comments have been addressed. Good call on checking the notebooks -- three of them would have been broken :) Also, the MDGXS notebooks are broken for me, but that is true even on develop, i.e., not as a result of this PR. MDGXS parts one and two both fail with on cell [22].
